File: Solutions/Cpp/two_sum_ii_input_array_is_sorted.cpp
/*
    [MEDIUM]
    167. Two Sum II - Input Array Is Sorted

    Concepts:
    - array
    - two pointers

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 20.35 MB  [Beats 5.64%]
*/

class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int n = numbers.size();
        int left = 0, right = n-1;

        std::vector<int> nums = numbers;
        for (int i = 0; i < nums.size(); i++)
            nums[i] += numbers[0];

        target += *std::min_element(nums.begin(), nums.end());

        while (left < right) {
            if (nums[left] + nums[right] > target)
                right--;
            else if (nums[left] + nums[right] < target)
                left++;
            else
                break;
        }
        return {left + 1, right + 1};
    }
};

File: Solutions/Rust/two_sum_ii_input_array_is_sorted.rs
/*
    [MEDIUM]
    167. Two Sum II - Input Array Is Sorted

    Concepts:
    - array
    - two pointers

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 2.48 MB   [Beats 2.63%]
*/
impl Solution {
    pub fn two_sum(numbers: Vec<i32>, mut target: i32) -> Vec<i32> {
        let n = numbers.len();
        let [mut left, mut right] = [0, n - 1];

        let mut nums = numbers.clone();
        for i in 0..n { nums[i] += numbers[0]; }

        target += *nums.iter().min().unwrap();

        while left < right {
            if nums[left] + nums[right] > target {
                right -= 1;
            } else if nums[left] + nums[right] < target {
                left += 1;
            } else {
                break;
            }
        }

        return vec![left as i32 + 1, right as i32 + 1];
    }
}